src/Umbraco.Core/Builder/UmbracoBuilder.cs

```diff
@@ -0,0 +1,72 @@
+﻿using Microsoft.Extensions.Configuration;
+using Microsoft.Extensions.DependencyInjection;
+using System;
+using System.Collections.Generic;
+using Umbraco.Core.Builder;
+using Umbraco.Core.Composing;
+
+namespace Umbraco.Web.Common.Builder
+{
+    public class UmbracoBuilder : IUmbracoBuilder
+    {
+        private readonly Dictionary<string, Action> _registrations = new Dictionary<string, Action>();
+        private readonly Dictionary<Type, ICollectionBuilder> _builders = new Dictionary<Type, ICollectionBuilder>();
+
+        public UmbracoBuilder(IServiceCollection services, IConfiguration config)
+        {
+            Services = services;
+            Config = config;
+        }
+
+        public IServiceCollection Services { get; }
+        public IConfiguration Config { get; }
+
+        /// <remarks>
+        /// TODO: Remove setter
+        /// This should be a constructor parameter
+        /// Attempting to fix it now opens a huge can of worms around logging setup
+        /// &amp; use of IOptionsMoniker&lt;HostingSettings&gt; for AspNetCoreHostingEnvironment
+        /// </remarks>
+        public TypeLoader TypeLoader { get; set; }
+
+        public IUmbracoBuilder AddWith(string key, Action add)
+        {
+            if (_registrations.ContainsKey(key)) return this;
+            _registrations[key] = add;
+            return this;
+        }
+
+        /// <summary>
+        /// Gets a collection builder (and registers the collection).
+        /// </summary>
+        /// <typeparam name="TBuilder">The type of the collection builder.</typeparam>
+        /// <returns>The collection builder.</returns>
+        public TBuilder WithCollectionBuilder<TBuilder>()
+            where TBuilder : ICollectionBuilder, new()
+        {
+            var typeOfBuilder = typeof(TBuilder);
+
+            if (_builders.TryGetValue(typeOfBuilder, out var o))
+                return (TBuilder)o;
+
+            var builder = new TBuilder();
+            _builders[typeOfBuilder] = builder;
+            return builder;
+        }
+
+        public void Build()
+        {
+            foreach (var a in _registrations)
+                a.Value();
+
+            _registrations.Clear();
+
+            // TODO: We can compose composers here, we have a typeloader, no need to do it in CoreRuntimeBootstrapper.
```

Comment:
Yep and all `TypeLoader` should do is look for `IComposer` (_`IUmbracoComposer`?_) instances, initialize them, and pass the builder across.  Everything should be registered against `IUmbracoBuilder`'s `IServiceCollection` via those implementations - services, property converters, events, the lot.

However, this shouldn't happen during `Build()` but should occur immediately after core services are registered. This means subsequent registration of services can be explicitly defined in `Startup`.

`Build()` is really only required if we maintain private lists within `UmbracoBuilder` that we manipulate during registration and then register en-masse. It might be better to use register against the `IServiceCollection` directly so the method can be dropped.

I'm very confident a single generic method that can be written that allows this and could replace all collection builder types.

Reply:
Composers now composed by UmbracoBuilder extension.

The call isn't in the perfect place yet but it keeps test passing for now, can move later.

043ec2d